### PR TITLE
Add `dream.load_nexus`

### PR DESCRIPTION
**Pull request by SimonHeybrock**
@celinedurniak this replaces https://github.com/scipp/ess/pull/199. I have adapted it for the latest file I have (which has sorted detector numbers).

### CODE REVIEW

File: tests/dream/io_test.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+@pytest.fixture
+def filename():
+    return dream.data.get_path('DREAM_nexus_sorted-2023-12-07.nxs')
+
+
+@pytest.mark.filterwarnings("ignore:Failed to load /entry/instrument/monitor_bunker")
+@pytest.mark.filterwarnings("ignore:Failed to load /entry/instrument/monitor_cave")
+@pytest.mark.filterwarnings("ignore:Failed to load /entry/instrument/polarizer/rate")
+@pytest.mark.filterwarnings("ignore:Failed to load /entry/instrument/sans_detector")
```
Comment (jl-wynen) [inline tests/dream/io_test.py:16]: Are you planning to remove those filters when we have a complete file?

**Comment by SimonHeybrock** [comment]
> > (which has sorted detector numbers).
> 
> I didn't follow the discussion. Can we rely on them being sorted from now on?

We might, but we are not: Scipp will dynamically chose a faster or slower grouping algorithms, depending on whether they are sorted or not. There is no risk for correctness due to assumptions.

File: tests/dream/io_test.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    assert 'instrument' in dg
+    instr = dg['instrument']
+    for name in (
+        'mantle',
+        'endcap_backward',
+        'endcap_forward',
+        'high_resolution',
+        'sans',
+    ):
+        assert f'{name}_detector' in instr
```
Comment (jl-wynen) [inline tests/dream/io_test.py:28]: In the GEANT4 loader, those are called `name`, i.e., without `"_detector"`. We should use the same names in both and I'm in favour of the shorter names in the GEANT4 loader. What do you think?

File: tests/dream/io_test.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+@pytest.mark.filterwarnings("ignore:Failed to load /entry/instrument/monitor_cave")
+@pytest.mark.filterwarnings("ignore:Failed to load /entry/instrument/polarizer/rate")
+@pytest.mark.filterwarnings("ignore:Failed to load /entry/instrument/sans_detector")
+def test_load_nexus_with_pixel_shape(filename):
+    dg = dream.load_nexus(filename, load_pixel_shape=True)
+    assert 'instrument' in dg
+    instr = dg['instrument']
+    # sans_detector is not populated in the current files
+    for name in ('mantle', 'endcap_backward', 'endcap_forward', 'high_resolution'):
+        assert f'{name}_detector' in instr
```
Comment (jl-wynen) [inline tests/dream/io_test.py:75]:
```suggestion
```
Already tested in line below

File: src/ess/dream/io.py [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+    *,
+    load_pixel_shape: bool = False,
+    entry: str = 'entry',
+    fold_detectors: bool = True,
+) -> sc.DataGroup:
+    """
+    Load an unprocessed DREAM NeXus file.
+
+    See https://confluence.esss.lu.se/pages/viewpage.action?pageId=462000005
+    and the ICD DREAM interface specification for details.
```
Comment (jl-wynen) [inline src/ess/dream/io.py:22]: This isn't great. Are there publicly accessible resources about this? Can we make these public?

File: src/ess/dream/io.py [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+    - Mounting-unit, cassette, and counter roughly correspond to the azimuthal angle
+      in the mantle detector. However, counter is reversed in the current files. This
+      may also be the case in the other detectors.
+    - The endcap detectors have a irregular structure that cannot be fully folded.
+      This is not a problem but note again that the counter may be reversed. It is
+      currently not clear if this is a bug in the file.
+    - The high-resolution detector has a very odd numbering scheme. The SANS detector
+      is using the same, but is not populated in the current files. The scheme
+      attempts to follows some sort of physical ordering in space (x,y,z), but it
+      looks partially messed up.
```
Comment (jl-wynen) [inline src/ess/dream/io.py:35]: Should these be issues to help us follow up on them?

File: src/ess/dream/io.py [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+
+    Parameters
+    ----------
+    path:
+        Path to the NeXus file.
+    load_pixel_shape:
+        If True, load the pixel shape from the file's NXoff_geometry group. This is
+        often unused by would slow down file loading. Default is False.
```
Comment (jl-wynen) [inline src/ess/dream/io.py:43]:
```suggestion
        often unused and would slow down file loading. Default is False.
```

File: src/ess/dream/io.py [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+    instrument = dg['instrument']
+    mantle = instrument['mantle_detector']
+    mantle['mantle_event_data'] = mantle['mantle_event_data'].fold(
+        dim='detector_number',
+        sizes={
+            'wire': 32,
+            'mounting_unit': 5,
+            'cassette': 6,
+            'counter': 2,
+            'strip': 256,
```
Comment (jl-wynen) [inline src/ess/dream/io.py:73]: These names correspond to the ICD, not GEANT4 names on the confluence page. Are these the ones we want to go with? Should we rename the dims in the GEANT4 loader?

File: src/ess/dream/io.py [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+    """
+    definitions = snx.base_definitions()
+    if not load_pixel_shape:
+        definitions["NXdetector"] = FilteredDetector
+
+    with snx.File(path, definitions=definitions) as f:
+        dg = f[entry][()]
+    dg = snx.compute_positions(dg)
+    if not fold_detectors:
+        return dg
```
Comment (jl-wynen) [inline src/ess/dream/io.py:63]: For readability, can you turn the everything below this line into a separate function and call it here?

File: tests/dream/io_test.py (hunk 1 above)
Comment (jl-wynen) [reply tests/dream/io_test.py:16]: I wanted to know if these are permanent and a problem with the loader or just temporary until we get better files.

File: src/ess/dream/io.py (hunk 4 above)
Comment (SimonHeybrock) [reply src/ess/dream/io.py:22]: Why public? This is an ESS repo, so I think everyone who needs access can get to the info?

File: src/ess/dream/io.py (hunk 5 above)
Comment (SimonHeybrock) [reply src/ess/dream/io.py:35]: I am not sure. It depends on how we will handle versioning in the loader. Will we have to keep it working with the current files, if ECDC makes fixes/changes in the future?

File: src/ess/dream/io.py (hunk 4 above)
Comment (jl-wynen) [reply src/ess/dream/io.py:22]: I don't know how that will work for visitors. But fine.

File: src/ess/dream/io.py (hunk 5 above)
Comment (jl-wynen) [reply src/ess/dream/io.py:35]: Maybe not the current ones as they should be replaced by newer ones, I guess. But ultimately, we will have to keep supporting old ones.

File: src/ess/dream/io.py (hunk 7 above)
Comment (jl-wynen) [reply src/ess/dream/io.py:73]: ok

**Pull request merged**